### PR TITLE
fix: shapes aligned with ticks test

### LINKED ISSUE #493: Bug in Certificate Project D3 Bar Chart
#### Issue Description

The D3 Bar Chart example project [at codepen](https://codepen.io/freeCodeCamp/pen/GrZVaM) has an bug that the x-axis representing the data.date is wrong. As the date and x-axis tick is not aligned, and inconsistent throughout the x-axis. This should violate the 
>"User Story #10: The data-date attribute and its corresponding bar element should align with the corresponding value on the x-axis."

The code using a d3 linear scale in a range from smallest year to largest year, i.e. line 57-63
```js
 var yearsDigits = years.map(function(item) {
    return item.substring(0, 4);
  });

var xScale = d3.scaleLinear()
    .domain([d3.min(yearsDigits), d3.max(yearsDigits)])
    .range([0, width]);
```
For example, the last bar, 2015-Q3 ends exactly at 2015 tick,
image

1950-Q1 column is after 1950 x-axis tick, 1955-Q1 is at the middle of 1955 tick, 1960-Q1 is right after 1960 tick, as shown below,
image
image
image

### PR DESCRIPTION
**Pull request by ValeraS**
#### Pre-Submission Checklist

- [x] Your pull request targets the `master` branch of freeCodeCamp/testable-projects-fcc
- [x] Branch starts with either `fix/`, `feature/`, or `translate/` (e.g. `fix/signin-issue`)
- [x] You have only one commit (if not, [squash](http://forum.freecodecamp.com/t/how-to-squash-multiple-commits-into-one-with-git/13231) them into one commit).
- [x] Your changes have been tested either locally or using a newly created CDN based on your fork's testable-projects-fcc/build/bundle.js file
- [ ] Working changes have been added to the build by running `npm run build`

#### Type of Change

- [ ] Small bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds new functionality)
- [x] Breaking change (fix or feature that would change existing functionality)
- [ ] Add new translation (feature adding new translations)

#### Checklist:

- [x] Tested changes locally.

- [x] Closes currently open issue: Closes #493

#### Description

in https://codepen.io/freeCodeCamp/pen/GrZVaM, need to replace js code with that:
```javascript
const projectName = 'bar-chart';
localStorage.setItem('example_project', 'D3: Bar Chart');
// coded by @Christian-Paul

var yMargin = 40,
    width = 800,
    height = 400,
    barWidth = width/275;

var tooltip = d3.select(".visHolder").append("div")
  .attr("id", "tooltip")
  .style("opacity", 0);

var overlay = d3.select('.visHolder').append('div')
  .attr('class', 'overlay')
  .style('opacity', 0);

var svgContainer =  d3.select('.visHolder')
    .append('svg')
    .attr('width', width + 100)
    .attr('height', height + 60);

d3.json('https://raw.githubusercontent.com/FreeCodeCamp/ProjectReferenceData/master/GDP-data.json', function(err, data) {
  
  svgContainer.append('text')
    .attr('transform', 'rotate(-90)')
    .attr('x', -200)
    .attr('y', 80)
    .text('Gross Domestic Product');
  
  svgContainer.append('text')
    .attr('x', width/2 + 120)
    .attr('y', height + 50)
    .text('More Information: http://www.bea.gov/national/pdf/nipaguid.pdf')
    .attr('class', 'info');
  
  var years = data.data.map(function(item) {
    var quarter;
    var temp = item[0].substring(5, 7);
    
    if(temp === '01') {
      quarter = 'Q1';
    }
    else if (temp === '04'){
      quarter = 'Q2';
    }
    else if(temp === '07') {
      quarter = 'Q3';
    }
    else if(temp ==='10') {
      quarter = 'Q4';
    }

    return item[0].substring(0, 4) + ' ' + quarter
  });
  
  var yearsDate = data.data.map(function(item) {
    return new Date(item[0]);
  });

  var xMax = new Date(d3.max(yearsDate));
  xMax.setMonth(xMax.getMonth() + 3);
  var xScale = d3.scaleTime()
    .domain([d3.min(yearsDate), xMax])
    .range([0, width]);
  
  var xAxis = d3.axisBottom()
    .scale(xScale);
  
  var xAxisGroup = svgContainer.append('g')
    .call(xAxis)
    .attr('id', 'x-axis')
    .attr('transform', 'translate(60, 400)');
  
  var GDP = data.data.map(function(item) {
    return item[1]
  });
  
  var scaledGDP = [];
  
  var gdpMin = d3.min(GDP);
  var gdpMax = d3.max(GDP);
  
  var linearScale = d3.scaleLinear()
    .domain([0, gdpMax])
    .range([0, height]);
  
  scaledGDP = GDP.map(function(item) {
    return linearScale(item);
  });
  
  var yAxisScale = d3.scaleLinear()
    .domain([0, gdpMax])
    .range([height, 0]);
  
  var yAxis = d3.axisLeft(yAxisScale)
    
  var yAxisGroup = svgContainer.append('g')
    .call(yAxis)
    .attr('id', 'y-axis')
    .attr('transform', 'translate(60, 0)');
    
  d3.select('svg').selectAll('rect')
    .data(scaledGDP)
    .enter()
    .append('rect')
    .attr('data-date', function(d, i) {
      return data.data[i][0]
    })
    .attr('data-gdp', function(d, i) {
      return data.data[i][1]
    })
    .attr('class', 'bar')
    .attr('x', function(d, i) {
      return xScale(yearsDate[i]);
    })
    .attr('y', function(d, i) {
      return height - d;
    })
    .attr('width', barWidth)
    .attr('height', function(d) {
      return d;
    })
    .style('fill', '#33adff')
    .attr('transform', 'translate(60, 0)')
    .on('mouseover', function(d, i) {
      overlay.transition()
        .duration(0)
        .style('height', d + 'px')
        .style('width', barWidth + 'px')
        .style('opacity', .9)
        .style('left', (i * barWidth) + 0 + 'px')
        .style('top', height - d + 'px')
        .style('transform', 'translateX(60px)');
      tooltip.transition()
        .duration(200)
        .style('opacity', .9);
      tooltip.html(years[i] + '<br>' + '$' + GDP[i].toFixed(1).replace(/(\d)(?=(\d{3})+\.)/g, '$1,') + ' Billion')
        .attr('data-date', data.data[i][0])
        .style('left', (i * barWidth) + 30 + 'px')
        .style('top', height - 100 + 'px')
        .style('transform', 'translateX(60px)');
    })
    .on('mouseout', function(d) {
      tooltip.transition()
        .duration(200)
        .style('opacity', 0);
      overlay.transition()
        .duration(200)
        .style('opacity', 0);
    });

});
```

### CODE REVIEW

**Comment by ValeraS** [comment]
@tbushman could you review it, please?

File: src/utils/alignment-D3.js [hunk 1]
```diff
@@ -84,56 +55,59 @@ export function areShapesAlignedWithTicks(
...
+      let prevPos = tickPositions[prevTick];
+      let nextPos = tickPositions[nextTick];
 
-      // if shape is positioned between the two ticks plus or minus 11px
-      // TODO reduce to 2px after Bar Chart is fixed. A leeway is necessary
-      // for this code to work on all chart types.
-      if ( (pos >= beforeTickPos - 11 ) && ( pos <= afterTickPos + 11 ) ) {
+      // If shape is positioned between the two ticks plus or minus 2px
+      // A leeway is necessary for this code to work on all chart types.
+      if ( prevPos - 2 <= pos && pos <= nextPos + 2 ) {
```
Comment (tbushman) [inline src/utils/alignment-D3.js]: @tchaffee Can you please swap out the fCC Codepen code with that provided by @ValeraS in this PR? Either that or this needs to be changed to 11 to get the official pen to pass. Additionally, I ran the tests against a few other projects and found that 3px was needed, if that's ok with you to change, @ValeraS  .

Comment (ValeraS) [reply src/utils/alignment-D3.js]: @tbushman Can I see these projects?

Comment (tbushman) [reply src/utils/alignment-D3.js]: https://codepen.io/deepikag/pen/KeGXLN

Comment (ValeraS) [reply src/utils/alignment-D3.js]: @tbushman I have gotten next results:
- for the first pen, all tests pass in my environment. (max variation - 0.5)
- the second one is buggy and doesn't pass the tests with a current CDN bundle. After fixing xScale, the alignment test passes. (max variation - 0)

to calculate max variation:
```javascript
      if (pos < prevPos) {
        maxVariation = Math.max(maxVariation, prevPos - pos);
      } else if (pos > nextPos) {
        maxVariation = Math.max(maxVariation, pos - nextPos);
      }
```

Comment (tbushman) [reply src/utils/alignment-D3.js]: @ValeraS Thanks for checking. The first linked Codepen above is the only one not passing -- in Firefox. It passes with 3px. Thanks.

**Comment by tchaffee** [comment]
I replaced the JS for the CodePen as requested. 

The backup CodePen with the old JS is here just in case: https://codepen.io/freeCodeCamp/pen/pxNYRg

**Comment by tchaffee** [comment]
@tbushman All check passed. I'll leave it up to you to decide whether or not to merge since you did the code review. Thanks!

File: src/utils/alignment-D3.js (hunk 1 above)
Comment (ValeraS) [reply src/utils/alignment-D3.js]: I can't reproduce this issue, but I have increased a lag.

**Comment by tbushman** [comment]
LGTM! Thanks for all your work, @ValeraS

**Pull request merged**